File: 2019/4/lib/aoc/parser.ex
defmodule AOC.Parser do
  def parse do
    {:ok, binary} = File.read("./input.txt")
    _parse(binary)
  end

  def parse(binary) do
    _parse(binary)
  end

  defp _parse(binary) do
    binary
    |> String.trim
    |> String.split("-")
    |> make_range
  end

  def make_range([low, high]) do
    (String.to_integer(low)..String.to_integer(high))
  end
end

File: 2020/2/lib/parser.rb
# frozen_string_literal: true

class Parser
  PASS_REGEX = /(?<low>\d+)\-(?<high>\d+) (?<letter>[a-z]): (?<password>[a-z]+)/.freeze

  def self.parse(input = nil)
    input ||= File.read(File.join(File.dirname(__FILE__), "../input.txt"))

    input
      .strip
      .split("\n")
      .map { |l| parse_line(l) }
  end

  def self.parse_line(line)
    Password.new(PASS_REGEX.match(line))
  end
end

class Password
  def initialize(match)
    @low = match["low"].to_i
    @high = match["high"].to_i
    @letter = match["letter"]
    @password = match["password"]
  end

  def valid?
    password.count(letter).between?(low, high)
  end

  def valid2?
    [password[low - 1], password[high - 1]].count(letter) == 1
  end

  private

  attr_reader :low, :high, :letter, :password
end
